### ISSUE TITLE
Taskset dependencies

### ISSUE THREAD

**Issue by boxerab** [dougbinks/enkiTS#46]
May I ask if you are planning on adding events in the near future? If you are, I would really like to hear about how you plan on designing them. If not, I would be interested in adding them as a PR. 

My motivation is: I am designing a completely asynch image decompressor, so I need to connect up different task sets with events, so that the completion of one task set causes a waiting set to enqueue itself. 

Interested to hear your plans for this feature.

**Comment by dougbinks** [comment]
I am planning on adding explicit dependencies, which I think would fill the role of what you call events (or continuations).

Due to the no-allocations and lock free approach of enkiTS there are a few issues in designing this, but I think I now have a solution which should work, although there is a small knock-on performance penalty.

In the meantime the way to achieve this with enkiTS is to wrap your taskset in another taskset, similarly to how the [ParallelReductionSumTaskSet works](https://github.com/dougbinks/enkiTS/blob/master/example/ParallelSum.cpp#L76-L96). You launch the image decompression from one taskset and then immediately wait on it, then when it completes launch the next task.

**Comment by boxerab** [comment]
Thanks, looking forward to these changes. It's a shame there is a performance penalty - hopefully it is possible to reduce the penalty by sacrificing some features ? As for wrapping taskset in another taskset, unfortunately this won't work for my use case as there will be multiple dependencies between tasksets.

**Comment by dougbinks** [comment]
If you have multiple dependencies then you can use multiple waits - care needs to be taken to be sure that you only wait on descendants (otherwise you can get a deadlock) but this works, I use it frequently.

**Comment by boxerab** [comment]
Interesting, thanks.

**Comment by dougbinks** [comment]
I've pushed the branch https://github.com/dougbinks/enkiTS/tree/dev_dependencies with a first, very rough draft, implementation of dependencies. There's no C interface as yet.

There's a new example showing how to use these:
https://github.com/dougbinks/enkiTS/blob/dev_dependencies/example/Dependencies.cpp

There's some work to go in improving the interface, avoid needless work, and ensure it's easy to use correctly and hard to use wrongly.

Feedback welcome.

**Comment by dougbinks** [comment]
There's a bug in the logic of the current code, I've not seen it manifest but I'd certainly not use this branch for anything serious. I have a fix in progress.

**Comment by dougbinks** [comment]
The [latest commit](https://github.com/dougbinks/enkiTS/commit/ba58dac7c94e706d7696c4b10e01ba4e14af7501) resolves the above issue and adds a new approach where you can now wait on an `enki::ICompletable` which is useful when you want to wait on multiple tasks completing, example in Dependencies.cpp

**Comment by dougbinks** [comment]
I've improved the dependencies example to show how to hide overhead in adding a complex task graph to enkiTS.

Whilst task graphs can be constructed once, reused many times, there is still overhead when adding the first task as dependencies need to be marked as not complete, so launching the task graph from another task is a good way to hide this.

https://github.com/dougbinks/enkiTS/commit/c1a0039fa6d9f6b2589d83319ce07f322e84e961

**Comment by dougbinks** [comment]
The latest update adds helper functions to ICompletable to make it easier to set dependencies, and support for using a vector to hold the dependencies.

The helper functions are a little uglier than I'd like, but without including <array> and <vector> to define specializations they're as simple as I think I can get.

In addition to some more testing to do, and adding some dtor code to ICompletable, I might switch the order of parameters to SetDependency() and perhaps rename pContinuationTask to pTaskToRunOnCompletion or similar.

**Comment by dougbinks** [comment]
Testing using our game Avoyd has been going well, and I've not encountered any issues in enkiTS depednecies for long runs of the stress testing (video of one run attached): https://www.youtube.com/watch?v=kLNhYXmd-Gk

**Comment by dougbinks** [comment]
This is now the last chance for feedback before I merge to master and release over the weekend.

I have now pushed changes which include the C API and C example to the https://github.com/dougbinks/enkiTS/tree/dev_dependencies branch.

**Comment by dougbinks** [comment]
This has now been merged to master and [released as v1.8](https://github.com/dougbinks/enkiTS/releases/tag/v1.8).
